File: ex4_2/java/src/Main.java
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Vector;

public class Main {

    private static <E> void startFinder(Collection<E> collection, IMatcher<E> matcher) {
        long start = System.nanoTime();
        E elem = Finder.findIf(collection.iterator(), matcher);
        long stop = System.nanoTime();

        double time = (stop - start)/1000000000.0;
        DecimalFormat f = new DecimalFormat("0.00");

        if (elem != null) {
            System.out.println("Found match in " + f.format(time) + " seconds.");
        }
    }

    public static void main(String[] args) {

        int size = 100000000;

        Vector<Integer> collection = new Vector<>();

        for (int i = 0; i < size-1; i++) {
            collection.add(1);
        }
        collection.add(0);

        for (int i = 0; i < 10; i++) {
            startFinder(collection, elem -> elem == 0);
        }
    }
}

File: ex4_2/cpp/main.cpp
#include <algorithm>
#include <iostream>
#include <iomanip>
#include <memory>
#include <string>
#include <vector>
#include <list>
#include <chrono>

using namespace std;

template<typename E, typename C>
void startFinder(C c, bool (*matcher)(const E)) {
    auto start = chrono::high_resolution_clock::now();
    auto elem_it = find_if(c.begin(), c.end(), matcher);
    auto stop = chrono::high_resolution_clock::now();

    auto time = chrono::duration_cast<chrono::nanoseconds>(stop - start).count()/1000000000.0;
    cout << setprecision(2) << fixed;

    if (elem_it != c.end()) {
        cout << "Found match in " << time << " seconds." << endl;
    }
}

int main() {

    const int size = 100000000;

    vector<int> collection;

    for (int i = 0; i < size-1; ++i) {
        collection.push_back(1);
    }
    collection.push_back(0);

    for (int i = 0; i < 10; ++i) {
        startFinder<int>(collection, [](int elem) {return elem == 0;});
    }

    return 0;
}